src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -1212,7 +1211,8 @@ public Boolean visitClassType(ClassType t, Type s) {
                 // If t is an intersection, sup might not be a class type
                 if (!sup.hasTag(CLASS)) return isSubtypeNoCapture(sup, s);
                 return sup.tsym == s.tsym
-                    && (t.tsym != s.tsym || t.isReferenceProjection() == s.isReferenceProjection())
+                    && (t.tsym != s.tsym ||
+                        (t.isReferenceProjection() == s.isReferenceProjection() && t.isValueProjection() == s.isValueProjection()))
```

Comment:
I guess this works because `asSuper` will return either:

* a supertype of `t`, in which case we now it's a reference type, so for `t` to be subtype of `s`, it must be that `t` is a reference type too
* it returns `t` itself, in which case subtyping holds as long as `t` and `s` are the same

Reply:
Right.